src/filter.rs

```diff
@@ -0,0 +1,300 @@
+use libc::{c_int, c_char, c_void, size_t};
+use std::{mem, slice, ptr};
+use {raw, call, util, Error};
+use std::ffi::{CString, CStr};
+use std::path::PathBuf;
+use raw::{git_filter, git_writestream};
+
+
+#[repr(C)]
+struct Filter<T> {
+    filter: git_filter,
+    instance: T,
+}
+
+#[repr(C)]
+struct Stream<T> {
+    parent: git_writestream,
+    handler: T,
+}
+
+/// Wrapper of the raw git_writestream object, representing the input
+/// of next filter of the chain
+pub struct RawStream {
+    stream: *mut git_writestream,
+}
+
+/// A trait representing operations to handle filter data stream.
+///
+/// These callbacks are typically called by libgit2.
+pub trait WriteStream {
```

Comment:
Let's have this inherit from `Send + 'static` for now, although we may be able to relax those in the future. I'd want to double check that libgit2 matches the ownership model we've got in the usage of a stream as well.